File: persistance/src/main/java/ro/mpp2024/ClientRepository.java
package ro.mpp2024;


import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Properties;

public class ClientRepository extends AbstractRepository<Integer, Client> implements IClientRepository {
//    public ClientRepository() {
//        super();
//    }

    public ClientRepository(Properties props) {
        super(props);
        logger.info("Initializing ClientRepository with properties: {} ",props);
    }

    @Override
    public Optional<Client> findById(Integer id) {
        logger.info("Find Client by ID: {}", id);
        String query = "SELECT * FROM Client WHERE id = ?";

        try (Connection connection = jdbc.getConnection();
             PreparedStatement statement = connection.prepareStatement(query)) {

            statement.setInt(1, id);
            ResultSet resultSet = statement.executeQuery();

            if (resultSet.next()) {
                String name = resultSet.getString("name");
                return Optional.of(new Client(id, name));
            }
        } catch (SQLException e) {
            logger.error("Database error while finding Client with id {}", id, e);
        }
        return Optional.empty();
    }

    @Override
    public Optional<Client> findByUsername(String username) {
        logger.info("Find Client by username: {}", username);
        String query = "SELECT * FROM Client WHERE name = ?";

        try (Connection connection = jdbc.getConnection();
             PreparedStatement statement = connection.prepareStatement(query)) {

            statement.setString(1, username);
            ResultSet resultSet = statement.executeQuery();

            if (resultSet.next()) {
                Integer id = resultSet.getInt("id");
                return Optional.of(new Client(id, username));
            }
        } catch (SQLException e) {
            logger.error("Database error while finding Client with username {}", username, e);
        }
        return Optional.empty();
    }

    public Optional<Client> findByName(String name) {
        logger.info("Find Client by name: {}", name);
        String query = "SELECT * FROM Client WHERE name = ?";

        try (Connection connection = jdbc.getConnection();
             PreparedStatement statement = connection.prepareStatement(query)) {

            statement.setString(1, name);
            ResultSet resultSet = statement.executeQuery();

            if (resultSet.next()) {
                return Optional.of(new Client(resultSet.getInt("id"), resultSet.getString("name")));
            }
        } catch (SQLException e) {
            logger.error("Database error while finding Client with name {}", name, e);
        }
        return Optional.empty();
    }

    @Override
    public Iterable<Client> findAll() {
        logger.info("Find all Clients");
        List<Client> clients = new ArrayList<>();
        String query = "SELECT * FROM Client";

        try (Connection connection = jdbc.getConnection();
             Statement statement = connection.createStatement();
             ResultSet resultSet = statement.executeQuery(query)) {

            while (resultSet.next()) {
                Integer id = resultSet.getInt("id");
                String name = resultSet.getString("name");
                clients.add(new Client(id, name));
            }
        } catch (SQLException e) {
            logger.error("Database error while finding all Clients", e);
        }
        return clients;
    }

    @Override
    public Optional<Client> save(Client client) {
        logger.info("Save Client: {}", client);
        String query = "INSERT INTO Client (name) VALUES (?)";

        try (Connection connection = jdbc.getConnection();
             PreparedStatement statement = connection.prepareStatement(query, Statement.RETURN_GENERATED_KEYS)) {

            statement.setString(1, client.getName());
            statement.executeUpdate();
            ResultSet resultSet = statement.getGeneratedKeys();

            if (resultSet.next()) {
                Integer id = resultSet.getInt(1);
                return Optional.of(new Client(id, client.getName()));
            }
        } catch (SQLException e) {
            logger.error("Database error while saving Client {}", client, e);
        }
        return Optional.empty();
    }

    @Override
    public Optional<Client> delete(Integer id) {
        logger.info("Delete Client with ID: {}", id);
        String query = "DELETE FROM Client WHERE id = ?";

        try (Connection connection = jdbc.getConnection();
             PreparedStatement statement = connection.prepareStatement(query)) {

            statement.setInt(1, id);
            statement.executeUpdate();
            return Optional.of(new Client(id, null));
        } catch (SQLException e) {
            logger.error("Database error while deleting Client with id {}", id, e);
        }
        return Optional.empty();
    }

    @Override
    public Optional<Client> update(Client client) {
        logger.info("Update Client: {}", client);
        String query = "UPDATE Client SET name = ? WHERE id = ?";

        try (Connection connection = jdbc.getConnection();
             PreparedStatement statement = connection.prepareStatement(query)) {

            statement.setString(1, client.getName());
            statement.setInt(2, client.getId());
            statement.executeUpdate();
            return Optional.of(client);
        } catch (SQLException e) {
            logger.error("Database error while updating Client {}", client, e);
        }
        return Optional.empty();
    }
}

File: csharpServer/persistance/ClientRepository.cs
using System;
using System.Collections.Generic;
using System.Data.SqlClient;
using Avalonia.Data;
using Microsoft.Data.Sqlite;
using Microsoft.Extensions.Logging;
using model;

namespace persistance;

public class ClientRepository : AbstractRepository<int, Client>, IClientRepository
{
    public ClientRepository() : base()
    {
        // Constructor specific for ClientRepository, if needed
    }

    public override Optional<Client> FindById(int id)
    {
        logger.LogInformation("Find Client by ID: {Id}", id);

        const string query = "SELECT * FROM Client WHERE id = @id";
        try
        {
            using (var connection = jdbc.GetConnection())
            using (var command = new SqliteCommand(query, (SqliteConnection)connection))
            {
                command.Parameters.AddWithValue("@id", id);
                connection.Open();

                using (var reader = command.ExecuteReader())
                {
                    if (reader.Read())
                    {
                        var name = reader.GetString(reader.GetOrdinal("name"));
                        return new Optional<Client>(new Client(id, name));
                    }
                }
            }
        }
        catch (SqliteException e)
        {
            logger.LogError(e, "Database error while finding Client with ID {Id}", id);
        }

        return new Optional<Client>();
    }

    public Optional<Client> FindByUsername(string username)
    {
        logger.LogInformation("Find Client by username: {Username}", username);

        if (string.IsNullOrEmpty(username))
            throw new ArgumentException("Username must not be null or empty", nameof(username));

        const string query = "SELECT * FROM Client WHERE name = @username";
        try
        {
            using (var connection = jdbc.GetConnection())
            using (var command = new SqliteCommand(query, (SqliteConnection)connection))
            {
                command.Parameters.AddWithValue("@username", username);
                connection.Open();

                using (var reader = command.ExecuteReader())
                {
                    if (reader.Read())
                    {
                        var id = reader.GetInt32(reader.GetOrdinal("id"));
                        return new Optional<Client>(new Client(id, username));
                    }
                }
            }
        }
        catch (SqliteException e)
        {
            logger.LogError(e, "Database error while finding Client with username {Username}", username);
        }

        return new Optional<Client>();
    }

    public Client FindByName(string name)
    {
        logger.LogInformation("Find Client by name: {Name}", name);

        if (string.IsNullOrEmpty(name))
            throw new ArgumentException("Name must not be null or empty", nameof(name));

        const string query = "SELECT * FROM Client WHERE name = @name";
        try
        {
            using (var connection = jdbc.GetConnection())
            using (var command = new SqliteCommand(query, (SqliteConnection)connection))
            {
                command.Parameters.AddWithValue("@name", name);
                connection.Open();

                using (var reader = command.ExecuteReader())
                {
                    if (reader.Read())
                    {
                        var id = reader.GetInt32(reader.GetOrdinal("id"));
                        return new Client(id, name);
                    }
                }
            }
        }
        catch (SqliteException e)
        {
            logger.LogError(e, "Database error while finding Client with name {Name}", name);
        }

        return null;
    }

    public override IEnumerable<Client> FindAll()
    {
        logger.LogInformation("Find all Clients");

        var clients = new List<Client>();
        const string query = "SELECT * FROM Client";

        try
        {
            using (var connection = jdbc.GetConnection())
            using (var command = new SqliteCommand(query, (SqliteConnection)connection))
            {
                connection.Open();

                using (var reader = command.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        var id = reader.GetInt32(reader.GetOrdinal("id"));
                        var name = reader.GetString(reader.GetOrdinal("name"));
                        clients.Add(new Client(id, name));
                    }
                }
            }
        }
        catch (SqliteException e)
        {
            logger.LogError(e, "Database error while finding all Clients");
        }

        return clients;
    }

    public override Optional<Client> Save(Client client)
        {
            logger.LogInformation("Save Client: {Client}", client);

            if (client == null)
                throw new ArgumentNullException(nameof(client));

            const string query = "INSERT INTO Client (name) VALUES (@name) RETURNING Id;";
            try
            {
                using (var connection = jdbc.GetConnection())
                using (var command = new SqliteCommand(query, (SqliteConnection)connection))
                {
                    command.Parameters.AddWithValue("@name", client.Name);
                    connection.Open();
                    var id = Convert.ToInt32(command.ExecuteScalar());
                    return new Optional<Client>(new Client(id, client.Name));
                }
            }
            catch (SqliteException e)
            {
                logger.LogError(e, "Database error while saving Client: {Client}", client);
            }

            return new Optional<Client>();
        }
    
    public override Optional<Client> Delete(int id)
    {
        logger.LogInformation("Delete Client with ID: {Id}", id);

        var clientToDelete = FindById(id);
        if (!clientToDelete.HasValue)
        {
            return new Optional<Client>();
        }

        const string query = "DELETE FROM Client WHERE id = @id";
        try
        {
            using (var connection = jdbc.GetConnection())
            using (var command = new SqliteCommand(query, (SqliteConnection)connection))
            {
                command.Parameters.AddWithValue("@id", id);
                connection.Open();
                var affectedRows = command.ExecuteNonQuery();

                if (affectedRows > 0)
                {
                    return clientToDelete;
                }
            }
        }
        catch (SqliteException e)
        {
            logger.LogError(e, "Database error while deleting Client with ID {Id}", id);
        }

        return new Optional<Client>();
    }

    public override Optional<Client> Update(Client client)
    {
        logger.LogInformation("Update Client: {Client}", client);

        if (client == null)
            throw new ArgumentNullException(nameof(client));

        const string query = "UPDATE Client SET name = @name WHERE id = @id";
        try
        {
            using (var connection = jdbc.GetConnection())
            using (var command = new SqliteCommand(query, (SqliteConnection)connection))
            {
                command.Parameters.AddWithValue("@name", client.Name);
                command.Parameters.AddWithValue("@id", client.Id);
                connection.Open();
                var affectedRows = command.ExecuteNonQuery();

                if (affectedRows > 0)
                {
                    return new Optional<Client>(client);
                }
            }
        }
        catch (SqliteException e)
        {
            logger.LogError(e, "Database error while updating Client: {Client}", client);
        }

        return new Optional<Client>();
    }
}